cgrok/tree.c

```diff
@@ -0,0 +1,198 @@
+#include "tree.h"
+#include <stdint.h>
+#include <string.h>
+
+// A shim to use dictlib trees instead of TC
+struct TCTREE {
+  dict *dict;
+  dict_itor *iter;
+};
+
+// Toyko cabinet accepts any arbitrary-sized data as a tree key.
+// We support this by using the first 4 bytes of the key to store the length.
+// So we have to wrap the libdict comparators with implementations that skip the first four bytes
+// of the key
+int tccmpint32(const void* k1, const void* k2) {
+  return dict_int_cmp(k1+4, k2+4);
+}
+
+int dict_var_str_cmp(const void* k1, const void* k2) {
+  return dict_str_cmp(k1+4, k2+4);
+}
+
+void tcfree(void *key, void *value) {
+  free(key);
+  free(value);
+}
+
+TCTREE *tctreenew(void) {
+  return tctreenew2(dict_var_str_cmp, 0);
+}
+
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop) {
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  tree->dict = hb_dict_new(cp, tcfree);
+  tree->iter = NULL;
+  return tree; 
+}
+
+// Iterate over elements in ascending key order
+void tctreeiterinit(TCTREE *tree) {
+  if (tree->iter != NULL) {
+    dict_itor_free(tree->iter);
+  }
+  tree->iter = dict_itor_new(tree->dict);
+  dict_itor_first(tree->iter);
+}
+
+const void *tctreeiternext(TCTREE *tree, int *sp) {
+  void *key = dict_itor_key(tree->iter);
+  dict_itor_next(tree->iter);
+  *sp = 0;
+  if (key) {
+    *sp = *(uint32_t*)key;
+    return key+4;
+  }
+  return NULL;
+}
+
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  void *val = tcpack(vbuf, vsiz);
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
```

Comment:
Should you check anything on the `inserted` value after doing an insert for possible failures? I'm assuming that's what its used for.